sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultVolumeSpec.java

```diff
@@ -25,33 +27,43 @@
      */
     private static final Pattern VALID_CONTAINER_PATH_PATTERN = Pattern.compile("[a-zA-Z0-9]+([a-zA-Z0-9_-]*)*");
 
+    /**
+     * Limit the length and characters in a profile name. A profile name should consist of alphanumeric
+     * characters ([a-zA-Z0-9]), dashes (-), underscores (_) or dots(.), and should be non-empty and at most 128
+     * characters. See: https://jira.mesosphere.com/browse/DCOS-40365
+     */
+    private static final Pattern VALID_PROFILE_PATTERN = Pattern.compile("[a-zA-Z0-9_.-]{1,128}");
+
     private final Type type;
     private final String containerPath;
+    private final List<String> profiles;
 
     public DefaultVolumeSpec(
```

Comment:
Since there are now different sets of arguments between ROOT/PATH vs MOUNT volumes, how about having two static methods here, one for MOUNT (with a profiles list) and another for ROOT/PATH (omits profiles list)?

Something like:
```
public static DefaultVolumeSpec createMount(... common args ..., List<String> profiles);
public static DefaultVolumeSpec createRootOrMount(Type type, ... common args ...);
```

Reply:
These two functions look a bit confusing, and the names won't be future-proof if we're going o support new volume types, such as PATH or BLOCK. If we want to have these new methods, it seems to me that the following provides a cleaner API:
```java
public static DefaultVolumeSpec createRootVolume(...);
public static DefaultVolumeSpec createMountVolume(...);
public static DefaultVolumeSpec updateVolume(VolumeSpec original, double diskSize);
```
However, this also means we would have the following code snippets:
`ResourceBuilder.java`:
```java
return resource.getDisk.hasSource()
        ? DefaultVolumeSpec.createMountVolume(..., Array.asList(profile), ...)
        : DefaultVolumeSpec.createRootVolume(...);
```
`DefaultResourceSet.java`:
```java
public Builder addRootVolume(...);
public Builder addMountVolume(...);
```
`YAMLToInternalMappers.java`:
```java
return volumeTypeEnum == VolumeSpec.Type.MOUNT
        ? DefaultVolumeSpec.createMountVolume(...)
        : DefaultVolumeSpec.createRootVolume(...);

if (rawVolume.getType() == VolumeSpec.Type.MOUNT.name()) {
    resourceSetBuilder.addMountVolume(...);
} else {
    resourceSetBuilder.addRootVolume(...);
}
```
I'm not sure if this change would lead to a cleaner API. WDYT?

Reply:
What you suggested looks great to me! Given that each of these types ends up being evaluated differently it's probably better to have explicit constructors for each.